src/flint/types/fmpq_mpoly.pyx

```diff
@@ -0,0 +1,600 @@
+from cpython.dict cimport PyDict_Size, PyDict_Check, PyDict_Next
+from cpython.tuple cimport PyTuple_Check, PyTuple_GET_SIZE
+
+from flint.types.fmpq cimport any_as_fmpq, fmpq, fmpq_set_any_ref
+from flint.types.fmpz cimport fmpz, fmpz_set_any_ref, any_as_fmpz
+from flint.types.fmpz_mpoly cimport fmpz_mpoly, fmpz_mpoly_ctx
+
+from flint.utils.typecheck cimport typecheck
+from flint.utils.conversion cimport str_from_chars
+
+from flint.flint_base.flint_base cimport flint_mpoly_context
+
+from flint.flintlib.flint cimport FMPZ_UNKNOWN
+from flint.flintlib.fmpq cimport fmpq_init, fmpq_clear, fmpq_is_zero, fmpq_set, fmpq_one
+from flint.flintlib.fmpz cimport fmpz_clear, fmpz_init, fmpz_set
+from flint.flintlib.fmpz_mpoly cimport fmpz_mpoly_set
+from flint.flintlib.fmpq_mpoly_factor cimport fmpq_mpoly_factor_t, \
+    fmpq_mpoly_factor_init, fmpq_mpoly_factor, fmpq_mpoly_factor_clear, \
+    fmpq_mpoly_factor_squarefree
+
+cdef extern from *:
+    """
+    /* An ugly hack to get around the ugly hack of renaming fmpq to avoid a c/python name collision */
+    typedef fmpq fmpq_struct;
+    """
+
+
+cimport cython
+cimport libc.stdlib
+
+cdef dict _fmpq_mpoly_ctx_cache = {}
+
+
+@cython.auto_pickle(False)
+cdef class fmpq_mpoly_ctx(flint_mpoly_context):
+    """
+    A class for storing the polynomial context
+
+    :param nvars: The number of variables in the ring
+    :param ordering:  The term order for the ring
+    :param names:  A tuple containing the names of the variables of the ring.
+
+    Do not construct one of these directly, use `fmpz_mpoly_ctx.get_context`.
+    """
+
+    _ctx_cache = _fmpq_mpoly_ctx_cache
+
+    def __init__(self, slong nvars, ordering, names):
+        if ordering == "lex":
+            fmpq_mpoly_ctx_init(self.val, nvars, ordering_t.ORD_LEX)
+        elif ordering == "deglex":
+            fmpq_mpoly_ctx_init(self.val, nvars, ordering_t.ORD_DEGLEX)
+        elif ordering == "degrevlex":
+            fmpq_mpoly_ctx_init(self.val, nvars, ordering_t.ORD_DEGREVLEX)
+        else:
+            raise ValueError("Unimplemented term order %s" % ordering)
+
+        super().__init__(nvars, names)
+
+    cpdef slong nvars(self):
+        """
+        Return the number of variables in the context
+
+            >>> ctx = fmpq_mpoly_ctx.get_context(4, "lex", 'x')
+            >>> ctx.nvars()
+            4
```

Comment:
Should this example not raise an error because there are too few names (`'x'`)?

Reply:
I believe this would produce variables like `x1`, `x2`, and so on.

Reply:
![image](https://github.com/flintlib/python-flint/assets/49361082/61620b88-87b2-4164-956c-918a82bda74d)
Yes this is fine, and functions as I suspected.